File: hello-edge-tts-python/multilingual_demo.py
#!/usr/bin/env python3
"""
Multilingual batch demo for Edge TTS Python implementation.
Generates audio files for 12 languages with specified sentences.
"""

import asyncio
import json
import os
import time
from pathlib import Path
import sys

# Add the parent directory to Python path to import local modules
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from tts_client import TTSClient, TTSError
from audio_player import AudioPlayer, AudioError
from utils import create_output_directory


async def load_language_config():
    """Load language configuration from shared config file."""
    config_path = "../shared/multilingual_demo_config.json"
    try:
        with open(config_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    except FileNotFoundError:
        print(f"Configuration file not found: {config_path}")
        return None
    except json.JSONDecodeError as e:
        print(f"Error parsing configuration file: {e}")
        return None


async def generate_audio_for_language(client, language_config, output_dir, play_audio=False):
    """Generate audio for a single language configuration."""
    lang_code = language_config['code']
    lang_name = language_config['name']
    flag = language_config['flag']
    text = language_config['text']
    voice = language_config['voice']
    alt_voice = language_config.get('alt_voice')
    
    print(f"\n{flag} {lang_name} ({lang_code.upper()})")
    print(f"Text: {text}")
    print(f"Voice: {voice}")
    
    try:
        # Try primary voice first
        audio_data = None
        used_voice = voice
        
        try:
            audio_data = await client.synthesize_text(text, voice)
        except TTSError as e:
            print(f"Primary voice failed: {e}")
            if alt_voice:
                print(f"Trying alternative voice: {alt_voice}")
                try:
                    audio_data = await client.synthesize_text(text, alt_voice)
                    used_voice = alt_voice
                except TTSError as e2:
                    print(f"Alternative voice also failed: {e2}")
                    raise e2
            else:
                raise e
        
        # Generate filename
        timestamp = int(time.time())
        lang_prefix = lang_code.split('-')[0]  # e.g., 'zh' from 'zh-cn'
        filename = f"multilingual_{lang_prefix}_python_{timestamp}.mp3"
        output_path = os.path.join(output_dir, filename)
        
        # Save audio
        await client.save_audio(audio_data, output_path)
        
        print(f"✅ Generated: {filename}")
        print(f"📁 Saved to: {output_path}")
        print(f"🎤 Used voice: {used_voice}")
        
        # Play audio if requested
        if play_audio:
            try:
                print("🔊 Playing audio...")
                player = AudioPlayer()
                player.play_file(output_path)
                print("✅ Playback completed")
            except AudioError as e:
                print(f"⚠️  Could not play audio: {e}")
        
        return True
        
    except Exception as e:
        print(f"❌ Failed to generate audio for {lang_name}: {e}")
        return False


async def main():
    """Main function for multilingual demo."""
    print("🌍 Multilingual Edge TTS Demo - Python Implementation")
    print("=" * 60)
    print("Generating audio for 12 languages with custom sentences...")
    
    # Load language configuration
    config = await load_language_config()
    if not config:
        print("❌ Failed to load language configuration")
        return 1
    
    languages = config.get('languages', [])
    if not languages:
        print("❌ No languages found in configuration")
        return 1
    
    print(f"📋 Found {len(languages)} languages to process")
    
    # Create output directory
    output_dir = "output"
    create_output_directory(output_dir)
    print(f"📁 Output directory: {os.path.abspath(output_dir)}")
    
    # Initialize TTS client
    try:
        client = TTSClient()
        print("✅ TTS client initialized")
    except Exception as e:
        print(f"❌ Failed to initialize TTS client: {e}")
        return 1
    
    # Process each language
    successful_count = 0
    failed_count = 0
    start_time = time.time()
    
    for i, language_config in enumerate(languages, 1):
        print(f"\n📍 Processing language {i}/{len(languages)}")
        
        success = await generate_audio_for_language(
            client, 
            language_config, 
            output_dir, 
            play_audio=False  # Set to True if you want to play each audio
        )
        
        if success:
            successful_count += 1
        else:
            failed_count += 1
        
        # Small delay between languages to be polite to the service
        if i < len(languages):
            print("⏳ Waiting before next language...")
            await asyncio.sleep(2)
    
    # Summary
    end_time = time.time()
    duration = end_time - start_time
    
    print(f"\n🏁 Processing Complete!")
    print("=" * 40)
    print(f"✅ Successful: {successful_count}")
    print(f"❌ Failed: {failed_count}")
    print(f"⏱️  Total time: {duration:.2f} seconds")
    print(f"📁 Output files saved in: {os.path.abspath(output_dir)}")
    
    if successful_count > 0:
        print(f"\n🎉 Successfully generated audio files for {successful_count} languages!")
        print("You can find all generated MP3 files in the output directory.")
    
    return 0 if failed_count == 0 else 1


if __name__ == "__main__":
    try:
        exit_code = asyncio.run(main())
        sys.exit(exit_code)
    except KeyboardInterrupt:
        print("\n⏹️  Operation cancelled by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n💥 Unexpected error: {e}")
        sys.exit(1)

File: hello-edge-tts-dart/bin/multilingual_demo.dart
#!/usr/bin/env dart

import 'dart:io';
import 'dart:convert';
import 'dart:typed_data';
import 'package:path/path.dart' as path;
import '../lib/hello_tts.dart';

/// Language configuration structure
class LanguageConfig {
  final String code;
  final String name;
  final String flag;
  final String text;
  final String voice;
  final String? altVoice;

  LanguageConfig({
    required this.code,
    required this.name,
    required this.flag,
    required this.text,
    required this.voice,
    this.altVoice,
  });

  factory LanguageConfig.fromJson(Map<String, dynamic> json) {
    return LanguageConfig(
      code: json['code'] as String,
      name: json['name'] as String,
      flag: json['flag'] as String,
      text: json['text'] as String,
      voice: json['voice'] as String,
      altVoice: json['alt_voice'] as String?,
    );
  }
}

/// Load language configuration from JSON file
Future<List<LanguageConfig>> loadLanguageConfig() async {
  const configPath = '../shared/multilingual_demo_config.json';
  
  try {
    final file = File(configPath);
    if (!await file.exists()) {
      print('❌ Configuration file not found: $configPath');
      return [];
    }

    final content = await file.readAsString();
    final Map<String, dynamic> config = json.decode(content);
    final List<dynamic> languagesJson = config['languages'] ?? [];

    return languagesJson
        .map((lang) => LanguageConfig.fromJson(lang as Map<String, dynamic>))
        .toList();
  } catch (e) {
    print('❌ Error loading configuration: $e');
    return [];
  }
}

/// Create output directory if it doesn't exist
Future<void> createOutputDirectory(String directory) async {
  final dir = Directory(directory);
  if (!await dir.exists()) {
    await dir.create(recursive: true);
  }
}

/// Generate audio for a single language
Future<bool> generateAudioForLanguage(
  TTSClient client,
  LanguageConfig languageConfig,
  String outputDir, {
  bool playAudio = false,
}) async {
  final langCode = languageConfig.code;
  final langName = languageConfig.name;
  final flag = languageConfig.flag;
  final text = languageConfig.text;
  final voice = languageConfig.voice;
  final altVoice = languageConfig.altVoice;

  print('\n$flag $langName (${langCode.toUpperCase()})');
  print('Text: $text');
  print('Voice: $voice');

  try {
    // Try primary voice first
    Uint8List? audioData;
    String usedVoice = voice;

    try {
      audioData = await client.synthesizeText(text, voice);
    } catch (e) {
      print('Primary voice failed: $e');
      if (altVoice != null) {
        print('Trying alternative voice: $altVoice');
        try {
          audioData = await client.synthesizeText(text, altVoice);
          usedVoice = altVoice;
        } catch (e2) {
          print('Alternative voice also failed: $e2');
          rethrow;
        }
      } else {
        rethrow;
      }
    }

    // Generate filename
    final timestamp = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    final langPrefix = langCode.split('-')[0]; // e.g., 'zh' from 'zh-cn'
    final filename = 'multilingual_${langPrefix}_dart_$timestamp.mp3';
    final outputPath = path.join(outputDir, filename);

    // Save audio
    final file = File(outputPath);
    await file.writeAsBytes(audioData);

    print('✅ Generated: $filename');
    print('📁 Saved to: $outputPath');
    print('🎤 Used voice: $usedVoice');

    // Play audio if requested
    if (playAudio) {
      try {
        print('🔊 Playing audio...');
        final player = AudioPlayer();
        await player.play(outputPath);
        print('✅ Playback completed');
      } catch (e) {
        print('⚠️  Could not play audio: $e');
      }
    }

    return true;
  } catch (e) {
    print('❌ Failed to generate audio for $langName: $e');
    return false;
  }
}

/// Main function for multilingual demo
Future<void> main() async {
  print('🌍 Multilingual Edge TTS Demo - Dart Implementation');
  print('=' * 60);
  print('Generating audio for 12 languages with custom sentences...');

  // Load language configuration
  final languages = await loadLanguageConfig();
  if (languages.isEmpty) {
    print('❌ Failed to load language configuration or no languages found');
    exit(1);
  }

  print('📋 Found ${languages.length} languages to process');

  // Create output directory
  const outputDir = 'output';
  await createOutputDirectory(outputDir);
  final outputPath = path.absolute(outputDir);
  print('📁 Output directory: $outputPath');

  // Initialize TTS client
  late TTSClient client;
  try {
    client = TTSClient();
    print('✅ TTS client initialized');
  } catch (e) {
    print('❌ Failed to initialize TTS client: $e');
    exit(1);
  }

  // Process each language
  int successfulCount = 0;
  int failedCount = 0;
  final startTime = DateTime.now();

  for (int i = 0; i < languages.length; i++) {
    final languageConfig = languages[i];
    print('\n📍 Processing language ${i + 1}/${languages.length}');

    final success = await generateAudioForLanguage(
      client,
      languageConfig,
      outputDir,
      playAudio: false, // Set to true if you want to play each audio
    );

    if (success) {
      successfulCount++;
    } else {
      failedCount++;
    }

    // Small delay between languages to be polite to the service
    if (i < languages.length - 1) {
      print('⏳ Waiting before next language...');
      await Future.delayed(const Duration(seconds: 2));
    }
  }

  // Summary
  final endTime = DateTime.now();
  final duration = endTime.difference(startTime);

  print('\n🏁 Processing Complete!');
  print('=' * 40);
  print('✅ Successful: $successfulCount');
  print('❌ Failed: $failedCount');
  print('⏱️  Total time: ${duration.inSeconds}.${duration.inMilliseconds % 1000} seconds');
  print('📁 Output files saved in: $outputPath');

  if (successfulCount > 0) {
    print('\n🎉 Successfully generated audio files for $successfulCount languages!');
    print('You can find all generated MP3 files in the output directory.');
  }

  exit(failedCount == 0 ? 0 : 1);
}